File: app/src/main/java/n7/mev/modules/ModulesViewModel.java
package n7.mev.modules;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.content.IntentSender;
import android.content.pm.PackageManager;
import android.widget.Toast;

import com.google.android.play.core.splitinstall.SplitInstallException;
import com.google.android.play.core.splitinstall.SplitInstallManager;
import com.google.android.play.core.splitinstall.SplitInstallManagerFactory;
import com.google.android.play.core.splitinstall.SplitInstallRequest;
import com.google.android.play.core.splitinstall.SplitInstallSessionState;
import com.google.android.play.core.splitinstall.SplitInstallStateUpdatedListener;
import com.google.android.play.core.splitinstall.model.SplitInstallErrorCode;
import com.google.android.play.core.splitinstall.model.SplitInstallSessionStatus;
import com.google.android.play.core.tasks.OnFailureListener;
import com.google.android.play.core.tasks.OnSuccessListener;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.databinding.Observable;
import androidx.databinding.ObservableBoolean;
import androidx.databinding.ObservableInt;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;
import n7.mev.R;
import n7.mev.util.SnackbarMessage;

public class ModulesViewModel extends AndroidViewModel implements SplitInstallStateUpdatedListener {

    public final SnackbarMessage showSnackbar = new SnackbarMessage();
    public final SnackbarMessage startConfirmationDialog = new SnackbarMessage();
    public final SplitInstallManager splitInstallManager;
    public ObservableBoolean showButtonAddModule = new ObservableBoolean(true);
    public MutableLiveData<Set<String>> availableModules = new MutableLiveData<>();
    public ObservableBoolean isLoading = new ObservableBoolean(false);
    public ObservableInt pbMaxValue = new ObservableInt();
    public ObservableInt pbProgressValue = new ObservableInt();
    public MutableLiveData<SplitInstallSessionState> splitInstallSessionState = new MutableLiveData<>();
    private Context context;
    private HashSet<String> allModules = new HashSet<>(Arrays.asList(
            "legion", "edi", "garrus", "grunt", "illusive_man", "jack", "avina",
            "jacob", "joker", "kasumi", "liara", "miranda", "mordin", "zaeed",
            "vega", "thane", "tali", "samara", "reaper", "javik","dlc_citadel"));

    public ModulesViewModel(@NonNull Application application) {
        super(application);
        this.context = application.getApplicationContext();
        splitInstallManager = SplitInstallManagerFactory.create(context);
        splitInstallManager.registerListener(this);
        updateModulesInSystem();
    }

    public void startConfirmationDialog(Activity activity) {
        try {
            splitInstallManager.startConfirmationDialogForResult(splitInstallSessionState.getValue(), activity, 0);
        } catch (IntentSender.SendIntentException e) {
            e.printStackTrace();
        }
    }

    public void updateModulesInSystem() {
        Set<String> installedModules = splitInstallManager.getInstalledModules();
        availableModules.setValue(installedModules);

        if (allModules.size() == installedModules.size()) {
            showButtonAddModule.set(false);
        } else {
            showButtonAddModule.set(true);
        }
    }

    public void deleteModule(String moduleName) {
        splitInstallManager.deferredUninstall(Collections.singletonList(moduleName))
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void aVoid) {
                        showSnackbar.setValue(R.string.deleted);
                    }
                });
        showSnackbar.setValue(R.string.delete_when_ready);
        updateModulesInSystem();
    }

    @SuppressWarnings("unchecked")
    public Set<String> modulesCanBeInstall() {
        Set<String> installedModules = splitInstallManager.getInstalledModules();
        Set<String> list = (Set<String>) allModules.clone();
        list.removeAll(installedModules);
        return list;
    }

    public void installModule(String moduleName) {
        if (!allModules.contains(moduleName)) return;

        isLoading.set(true);
        showButtonAddModule.set(false);

        SplitInstallRequest request = SplitInstallRequest.newBuilder()
                .addModule(moduleName)
                .build();
        splitInstallManager.startInstall(request)
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(Exception e) {
                        switch (((SplitInstallException) e).getErrorCode()) {
                            default:
                            case SplitInstallErrorCode.NETWORK_ERROR:
                                showSnackbar.setValue(R.string.network_error);
                                isLoading.set(false);
                                showButtonAddModule.set(true);
                                break;
                        }
                    }
                });
//                .addOnCompleteListener(new OnCompleteListener<Integer>() {
//                    @Override
//                    public void onComplete(Task<Integer> task) {
//                        isLoading.set(false);
//                        updateModulesInSystem();
//                    }
//                });
    }

    public void refreshContext() {
        try {
            context = context.createPackageContext(context.getPackageName(), 0);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onStateUpdate(SplitInstallSessionState splitInstallSessionState) {
        int status = splitInstallSessionState.status();
        switch (status) {
            case SplitInstallSessionStatus.REQUIRES_USER_CONFIRMATION:
                showSnackbar.setValue(R.string.require);
                this.splitInstallSessionState.setValue(splitInstallSessionState);
                startConfirmationDialog.call();
//                splitInstallSessionState.resolutionIntent();
                break;
            case SplitInstallSessionStatus.DOWNLOADING:
                displayLoadingState(splitInstallSessionState);
//                showSnackbar.setValue(R.string.downloading);
                isLoading.set(true);
                showButtonAddModule.set(false);
                break;
            case SplitInstallSessionStatus.INSTALLING:
                displayLoadingState(splitInstallSessionState);
                showSnackbar.setValue(R.string.installing);
                isLoading.set(true);
                showButtonAddModule.set(false);
                break;
            case SplitInstallSessionStatus.INSTALLED:
                showSnackbar.setValue(R.string.installed);
                isLoading.set(false);
                showButtonAddModule.set(true);
                updateModulesInSystem();
                break;
            case SplitInstallSessionStatus.FAILED:
                String error = splitInstallSessionState.errorCode() + " for module " + splitInstallSessionState.moduleNames();
                Toast.makeText(getApplication(), error, Toast.LENGTH_SHORT).show();
                break;
        }
    }

    private void displayLoadingState(SplitInstallSessionState state) {

        int max = Math.round(state.totalBytesToDownload());
        pbMaxValue.set(max);
        int progress = Math.round(state.bytesDownloaded());
        pbProgressValue.set(progress);

    }

    @Override
    protected void onCleared() {
        splitInstallManager.unregisterListener(this);
        super.onCleared();
    }
}

File: app/src/main/java/n7/mev/modules/ModulesViewModel.kt
package n7.mev.modules

import android.app.Activity
import android.app.Application
import android.content.Context
import android.content.IntentSender.SendIntentException
import android.content.pm.PackageManager
import android.widget.Toast
import androidx.databinding.ObservableBoolean
import androidx.databinding.ObservableInt
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.google.android.play.core.splitinstall.SplitInstallException
import com.google.android.play.core.splitinstall.SplitInstallManager
import com.google.android.play.core.splitinstall.SplitInstallManagerFactory
import com.google.android.play.core.splitinstall.SplitInstallRequest
import com.google.android.play.core.splitinstall.SplitInstallSessionState
import com.google.android.play.core.splitinstall.SplitInstallStateUpdatedListener
import com.google.android.play.core.splitinstall.model.SplitInstallErrorCode
import com.google.android.play.core.splitinstall.model.SplitInstallSessionStatus
import n7.mev.R
import n7.mev.data.source.local.FeatureModule

fun <T> MutableLiveData<T>.setSingleEvent(value: T) {
    this.value = value
    this.value = null
}

class ModulesViewModel(application: Application) : AndroidViewModel(application), SplitInstallStateUpdatedListener {

    companion object {
        const val PREFIX = "feature_"
    }

    private val _installedModules = MutableLiveData<List<FeatureModule>>()
    val installedModules: LiveData<List<FeatureModule>> = _installedModules

    private val _buttonVisibility = MutableLiveData<Boolean>()
    val buttonVisibility: LiveData<Boolean> = _buttonVisibility

    private val _showMessage = MutableLiveData<Int>()
    val showMessage: LiveData<Int?> = _showMessage

    private val _showConfirmationDialog = MutableLiveData<Boolean>()
    val showConfirmationDialog: LiveData<Boolean?> = _showConfirmationDialog

    private val splitInstallManager: SplitInstallManager = SplitInstallManagerFactory.create(application)


    @kotlin.jvm.JvmField
    var isLoading = ObservableBoolean(false)

    @kotlin.jvm.JvmField
    var pbMaxValue = ObservableInt()

    @kotlin.jvm.JvmField
    var pbProgressValue = ObservableInt()
    private var splitInstallSessionState = MutableLiveData<SplitInstallSessionState>()
    private var context: Context
    private val allModules = hashSetOf(
            "legion", "edi", "garrus", "grunt", "illusive_man", "jack", "feature_avina",
            "jacob", "joker", "kasumi", "liara", "miranda", "mordin", "zaeed",
            "vega", "thane", "tali", "samara", "reaper", "javik", "dlc_citadel")


    init {
        context = application.applicationContext
        splitInstallManager.registerListener(this)
        updateInstalledModules()
    }

    private fun updateInstalledModules() {
        val moduleList: List<FeatureModule> = splitInstallManager.installedModules.map {
            FeatureModule(it.replace(PREFIX, ""), it)
        }
        _installedModules.value = moduleList
        _buttonVisibility.value = allModules.size != _installedModules.value!!.size
    }

    fun startConfirmationDialog(activity: Activity) {
        try {
            splitInstallManager.startConfirmationDialogForResult(splitInstallSessionState.value, activity, 0)
        } catch (e: SendIntentException) {
            e.printStackTrace()
        }
    }

    fun deleteModule(moduleName: String) {
        splitInstallManager.deferredUninstall(listOf(moduleName))
                .addOnSuccessListener { _showMessage.value = R.string.delete_when_ready }
        updateInstalledModules()
    }

    fun installedModules(): List<FeatureModule> {
        val availableList = allModules - splitInstallManager.installedModules.toList()
        return availableList.map { FeatureModule(it.replace(PREFIX, ""), it) }
    }

    fun installModule(moduleName: String) {
        if (!allModules.contains(moduleName)) return
        isLoading.set(true)
        _buttonVisibility.setValue(false)
        val request = SplitInstallRequest.newBuilder()
                .addModule(moduleName)
                .build()
        splitInstallManager.startInstall(request)
                .addOnFailureListener { e ->
                    when ((e as SplitInstallException).errorCode) {
                        SplitInstallErrorCode.NETWORK_ERROR -> {
                            _showMessage.value = R.string.network_error
                            isLoading.set(false)
                            _buttonVisibility.setValue(true)
                        }
                        else -> {
                            _showMessage.value = R.string.network_error
                            isLoading.set(false)
                            _buttonVisibility.setValue(true)
                        }
                    }
                }
                .addOnSuccessListener {
                    isLoading.set(false)
                    updateInstalledModules()
                }
    }

    fun refreshContext() {
        try {
            context = context.createPackageContext(context.packageName, 0)
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
    }

    override fun onStateUpdate(splitInstallSessionState: SplitInstallSessionState) {
        val status = splitInstallSessionState.status()
        when (status) {
            SplitInstallSessionStatus.REQUIRES_USER_CONFIRMATION -> {
                _showMessage.setSingleEvent(R.string.require)
                this.splitInstallSessionState.value = splitInstallSessionState
                _showConfirmationDialog.setSingleEvent(true)
            }
            SplitInstallSessionStatus.DOWNLOADING -> {
                displayLoadingState(splitInstallSessionState)
                //                showSnackbar.setValue(R.string.downloading);
                isLoading.set(true)
                _buttonVisibility.setValue(false)
            }
            SplitInstallSessionStatus.INSTALLING -> {
                displayLoadingState(splitInstallSessionState)
                _showMessage.setSingleEvent(R.string.installing)
                isLoading.set(true)
                _buttonVisibility.setValue(false)
            }
            SplitInstallSessionStatus.INSTALLED -> {
                _showMessage.setSingleEvent(R.string.installed)
                isLoading.set(false)
                _buttonVisibility.setValue(true)
                updateInstalledModules()
            }
            SplitInstallSessionStatus.FAILED -> {
                val error = splitInstallSessionState.errorCode().toString() + " for module " + splitInstallSessionState.moduleNames()
                Toast.makeText(getApplication(), error, Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun displayLoadingState(state: SplitInstallSessionState) {
        val max = Math.round(state.totalBytesToDownload().toFloat())
        pbMaxValue.set(max)
        val progress = Math.round(state.bytesDownloaded().toFloat())
        pbProgressValue.set(progress)
    }

    override fun onCleared() {
        splitInstallManager.unregisterListener(this)
        super.onCleared()
    }

}